### ISSUE TITLE
`run_cmd` in `locals` block exectuted twice

### ISSUE THREAD

**Issue by bendavies** [gruntwork-io/terragrunt#1427]
Hi there,

It seems a `run_cmd` inside a `locals` block will be run twice.

### Reproducer:
`terragrunt.hcl`
```hcl
locals {
  foo = run_cmd("echo", uuid())
}
```

```bash
terragrunt -v
terragrunt version v0.26.2

terragrunt init
[terragrunt] 2020/11/13 15:33:15 Running command: echo d3d3c70d-4165-6842-f27a-1d846be06553
d3d3c70d-4165-6842-f27a-1d846be06553
[terragrunt] 2020/11/13 15:33:15 run_cmd output: [d3d3c70d-4165-6842-f27a-1d846be06553
]
[terragrunt] [/path] 2020/11/13 15:33:15 Running command: terraform --version
[terragrunt] 2020/11/13 15:33:16 Terraform version: 0.12.29
[terragrunt] 2020/11/13 15:33:16 Reading Terragrunt config file at /path/terragrunt.hcl
[terragrunt] 2020/11/13 15:33:16 Running command: echo e654adfb-84bc-3326-9b89-71198959deee
e654adfb-84bc-3326-9b89-71198959deee
[terragrunt] 2020/11/13 15:33:16 run_cmd output: [e654adfb-84bc-3326-9b89-71198959deee
]
[terragrunt] 2020/11/13 15:33:16 Running command: terraform init
Terraform initialized in an empty directory!

The directory has no Terraform configuration files. You may begin working
with Terraform immediately by creating Terraform configuration files.
```

notice two different UUIDs in the log, which I believe shows this is not just a double logging issue.

Thanks!

**Comment by ksvladimir** [comment]
If definitely happens to me too, I was about to report the same issue.

**Comment by brikis98** [comment]
Hm, must be an issue with the parsing logic, as it takes several passes to parse everything. If anyone has time to dig into this, help would be greatly appreciated, as we're quite booked right now.

**Comment by amnk** [comment]
This happens because of [checkVersionConstraints()](https://github.com/gruntwork-io/terragrunt/blob/master/cli/cli_app.go#L290), which leads to [cli_app.go#L405](https://github.com/gruntwork-io/terragrunt/blob/master/cli/cli_app.go#L405). 

Since [config.PartialParseConfigFile()](https://github.com/gruntwork-io/terragrunt/blob/master/config/config_partial.go#L165) always parses `locals` block, we get two executions.

**Comment by brikis98** [comment]
> This happens because of [checkVersionConstraints()](https://github.com/gruntwork-io/terragrunt/blob/master/cli/cli_app.go#L290), which leads to [cli_app.go#L405](https://github.com/gruntwork-io/terragrunt/blob/master/cli/cli_app.go#L405).
> 
> Since [config.PartialParseConfigFile()](https://github.com/gruntwork-io/terragrunt/blob/master/config/config_partial.go#L165) always parses `locals` block, we get two executions.

Oh, bummer. Any thoughts on how to fix this?

**Comment by tiwood** [comment]
@yorinasub17, in #1554 you had some great suggestions, do you already have a timeframe in which you plan to implement those? Is there any way one could help?

We're currently discovering a lot of dynamic information with a some `run_cmd` commands, which are quite expensive time-wise and the duplicate execution is not making it any better :-(

**Comment by ChristophShyper** [comment]
Should it be executed even when it's in negative `if` condition?
For config below it's executed always, and twice.
```
terraform {
  source = "/my_module"
}

inputs = {
  my_cmd = local.my_cmd
  my_value = read_terragrunt_config(find_in_parent_folders("my_config.hcl")).locals.my_value
}

locals {
  my_cmd = local.my_value == "some_value" ? jsondecode(run_cmd("--terragrunt-quiet", "python3", "./my_script.py")) : {}
}
```

When running without `--terragrunt-quiet` I see it's executed twice before `terraform init`.

**Comment by brikis98** [comment]
Fixed in #1763 and released in https://github.com/gruntwork-io/terragrunt/releases/tag/v0.31.6.

**Comment by Tzrlk** [comment]
I think this may have regressed. I'm running v0.36.7 and the script is executing several times without caching the output:
```
>terragrunt apply
Please enter the deployment type: test
Deployment type 'test' selected.
test
Please enter the deployment type: test
Deployment type 'test' selected.
test
Please enter the deployment type: test
Deployment type 'test' selected.
test
Please enter the deployment type: test
Deployment type 'test' selected.
test
Please enter the deployment type: test
Deployment type 'test' selected.
test
time=2022-07-15T12:58:14+12:00 level=info msg=Executing hook: AwsAuthCheck
```
```hcl
locals {
	config = yamldecode(file("${get_parent_terragrunt_dir()}/config.yml"))
	deploy_type = trimspace(lower(run_cmd("sh", "${get_parent_terragrunt_dir()}/../bin/get-deploy-type.sh")))
	# ...
```
The bash script just prompts for valid input and only prints to stdout at the very end. Once all invocations of the script succeed, everything works as expected. Happy to know if I've just missed something.
